src/components/DynamicDimension/DynamicDimension.js

```diff
@@ -1,139 +1,61 @@
-import React, { Component } from 'react'
+import React, { useEffect, useState } from 'react'
 import PropTypes from 'prop-types'
 
-import ItemSelector from '../ItemSelector/ItemSelector'
-import FilterField from '../FilterField'
+import ItemSelector from './ItemSelector'
 import { apiFetchItemsByDimension } from '../../api/dimensions'
 
-export class DynamicDimension extends Component {
-    state = {
-        filterText: '',
-        nextPage: null,
-        items: [],
-        unselectedIds: [],
-        selectedIds: [],
-    }
-
-    componentDidMount = async () => {
-        const items = await apiFetchItemsByDimension(
-            this.props.d2,
-            this.props.dialogId
-        )
-
-        this.setState({ items })
-    }
-
-    onClearFilter = () => {
-        this.setState({ filterText: '' })
-    }
-
-    onFilterTextChange = filterText => {
-        const filteredItems = this.state.items.map(
-            item =>
-                item.name.toLowerCase().includes(filterText.toLowerCase()) &&
-                item.id
-        )
-
-        this.setState({
-            filterText,
-            unselectedIds: filteredItems,
-        })
-    }
-
-    selectItems = selectedIds => {
-        const unselectedIds = this.state.unselectedIds.filter(
-            id => !selectedIds.includes(id)
-        )
-        this.setState({ unselectedIds })
-
-        const itemsToAdd = this.state.items.filter(di =>
-            selectedIds.includes(di.id)
-        )
-
-        this.props.onSelect({
-            dimensionId: this.props.dialogId,
-            items: [
-                ...this.props.selectedItems.filter(
-                    item => !selectedIds.includes(item.id)
-                ),
-                ...itemsToAdd,
-            ],
-        })
-    }
-
-    deselectItems = ids => {
-        const unselectedIds = [
-            ...new Set([...this.state.unselectedIds, ...ids]),
-        ]
-        this.setState({ unselectedIds })
-
-        this.props.onDeselect({
-            dimensionId: this.props.dialogId,
-            itemIdsToRemove: ids,
+export const DynamicDimension = ({
+    context,
+    dimensionId,
+    onSelect,
+    selectedItems,
+    rightFooter,
+}) => {
+    const [items, setItems] = useState([])
+
+    useEffect(() => {
+        if (!items || !items.length) {
+            getItems()
+        }
+    }, [])
+
+    const getItems = async () =>
+        setItems(await apiFetchItemsByDimension(context, dimensionId))
+    // TODO: This needs to be refactored to use a loading spinner once Transfer supports it: https://jira.dhis2.org/browse/TECH-379
```

Comment:
This can be easily achieved with the data engine.
When looking into adding the spinner I think we should also replace d2 with data engine.